TemperatureMeter.c

```diff
@@ -0,0 +1,153 @@
+#include <string.h>
+#include <limits.h>
+#include <errno.h>
+#include <sensors/sensors.h>
+
+typedef struct
+{
+   int size;
+   int* array;
+} Dynarr;
+
+static int temp_converter(Meter* this, double temp) {
+   if (this->pl->settings->fahrenheitTemperature)
+      return temp * 9 / 5 + 32;
+   else
+      return temp;
+}
+
+static Dynarr* get_cpu_temp(Meter* this, sensors_subfeature_type type) {
+   int i = 0, n = 0, count = 0;
+   Dynarr* result = malloc(sizeof(Dynarr));
+
+   for (const sensors_chip_name *chip = sensors_get_detected_chips(NULL, &n); chip; chip = sensors_get_detected_chips(NULL, &n)) {
+      char buffer[64];
+      if (sensors_snprintf_chip_name(buffer, sizeof(buffer), chip) < 0)
+         continue;
+
+      // cpu_thermal is for Raspberry Pi sensor support
+      if (!String_startsWith(buffer, "coretemp") && !String_startsWith(buffer, "cpu_thermal"))
+         continue;
+
+      int m = 0;
+      for (const sensors_feature *feature = sensors_get_features(chip, &m); feature; feature = sensors_get_features(chip, &m)) {
+         if (feature->type == SENSORS_FEATURE_TEMP)
+            count++;
+      }
+
+      result->array = calloc(count, sizeof(int));
+      result->size = count;
+
+      m = 0;
+      for (const sensors_feature *feature = sensors_get_features(chip, &m); feature; feature = sensors_get_features(chip, &m)) {
+         if (feature->type != SENSORS_FEATURE_TEMP)
+            continue;
+
+         const sensors_subfeature *sub_feature = sensors_get_subfeature(chip, feature, type);
+         if (sub_feature) {
+            double local_temp;
+            if (sensors_get_value(chip, sub_feature->number, &local_temp) == 0) {
+               result->array[i] = temp_converter(this, local_temp);
+               ++i;
+            }
+         }
+      }
+   }
+
+   return result;
+}
+
+int TemperatureMeter_attributes[] = {
+   TEMPERATURE_COOL,
+   TEMPERATURE_MEDIUM,
+   TEMPERATURE_HOT,
+};
+
+static void TemperatureMeter_updateValues(Meter* this, char* buffer, int len) {
+   Dynarr* core_temperature = get_cpu_temp(this, SENSORS_SUBFEATURE_TEMP_INPUT);
+   double temperature = core_temperature->array[0];
+
+   // Show only package temperature in bars meter mode
+   if (temperature < 0) {
+      xSnprintf(buffer, len, "N/A");
+      return;
+   }
+   this->total = get_cpu_temp(this, SENSORS_SUBFEATURE_TEMP_MAX)->array[0];
+
+   this->values[0] = temperature;
+   if (this->pl->settings->fahrenheitTemperature)
+      xSnprintf(buffer, len, "%d%sF", (int) temperature, CRT_degreeSign());
+   else
+      xSnprintf(buffer, len, "%d%sC", (int) temperature, CRT_degreeSign());
+   free(core_temperature);
+}
+
+static void TemperatureMeter_display(Object* cast, RichString* out) {
+   Meter* this                = (Meter*)cast;
+   Dynarr* cores_temperatures = get_cpu_temp(this, SENSORS_SUBFEATURE_TEMP_INPUT);
+   int temp_medium            = temp_converter(this, 60);
+   int temp_hot               = temp_converter(this, 70);
```

Comment:
Not sure what common values nowadays for these should be. But given we are talking about 10K of 70K effective range (35°C - 105°C), we probably should make this a bit wider: 50-55°C for medium, 75-85°C for hot?